### ISSUE TITLE
No TLS handshake after reconnect (syslog driver)

### ISSUE THREAD

**Issue by willcey** [syslog-ng/syslog-ng#3450]
Hello

I'm trying to get a syslog "proxy" running. It should receive syslog messages from a source (syslog driver with TLS) and immediately forward it to a destination (syslog driver with TLS).

Syslog-ng Server Version is 3.23.1 on RHEL 8. You will find my configuration attached. In the global config i have only the following set:
```
@version: 3.23
@include "scl.conf"
options {
    time_reopen (5);
};
@include "/etc/syslog-ng/conf.d/*.conf"
```

If i start my syslog-ng Server the server connects to the destination:
```
[2020-10-05T22:39:07.657443] Syslog connection established; fd='13', server='AF_INET(remote_dst_ip:dst_port)', local='AF_INET(syslog-ng_internal_ip:0)'
```

As soon as i'm sending a message from my client (log_client_ip) to the source (syslog-ng_public_ip) the following appears:
```
[2020-10-05T22:39:16.289593] Syslog connection accepted; fd='15', client='AF_INET(log_client_ip:10400)', local='AF_INET(syslog-ng_public_ip:6000)'
[2020-10-05T22:39:16.499684] Incoming log entry; line='<85>1 2020-10-05T20:39:15.531Z log_client_hostname IPF 12920 IHE+RFC-3881 - <?xml version="1.0" encoding="UTF-8"?>XML_Payload
[2020-10-05T22:39:16.501307] Outgoing message; message='<85>1 2020-10-05T20:39:15+00:00 log_client_ip IPF 12920 IHE+RFC-3881 - <?xml version="1.0" encoding="UTF-8"?>XML_Payload
[2020-10-05T22:39:16.501464] Can't send the message rewind backlog; message='<85>1 2020-10-05T20:39:15+00:00 log_client_ip IPF 12920 IHE+RFC-3881 - <?xml version="1.0" encoding="UTF-8"?>XML_Payload
[2020-10-05T22:39:16.504029] Syslog connection closed; fd='15', client='AF_INET(log_client_ip:10400)', local='AF_INET(syslog-ng_public_ip:6000)'
[2020-10-05T22:39:16.504646] Certificate validation progress; subject='CN=Root CA, OU=, O=Organization, C=CH', issuer='CN=Root CA, OU=, O=Organization, C=CH'
[2020-10-05T22:39:16.504752] Certificate validation progress; subject='CN=Intermediate CA, OU=, O=Organization, C=CH', issuer='CN=Root CA, OU=, O=Organization, C=CH'
[2020-10-05T22:39:16.504799] Certificate validation progress; subject='CN=remote_dst_host, OU=, O=Organization, L=Location, ST=State, C=CH', issuer='CN=Intermediate CA, OU=, O=Organization, C=CH'
[2020-10-05T22:39:16.504810] Certificate subject matches configured hostname; hostname='remote_dst_host', certificate='remote_dst_host'
[2020-10-05T22:39:16.506712] Outgoing message; message='<85>1 2020-10-05T20:39:15+00:00 log_client_ip IPF 12920 IHE+RFC-3881 - <?xml version="1.0" encoding="UTF-8"?>XML_Payload
```

Except for the message "Can't send the message rewind backlog" this looks pretty ok to me and i also receive the message correctly on the destination.

I can send multiple messages like this and it works. But as soon as i wait for 30 seconds the remote destination is closing the connection. At that point i have nothing in the Syslog-NG log.

After sending a new message to the source i see the following:
```
[2020-10-05T22:43:02.471451] Syslog connection accepted; fd='15', client='AF_INET(log_client_ip:10400)', local='AF_INET(syslog-ng_public_ip:6000)'
[2020-10-05T22:43:02.760481] Incoming log entry; line='<85>1 2020-10-05T20:43:01.728Z log_client_hostname IPF 18060 IHE+RFC-3881 - <?xml version="1.0" encoding="UTF-8"?>XML_Payload
[2020-10-05T22:43:02.762985] Syslog connection closed; fd='15', client='AF_INET(log_client_ip:10400)', local='AF_INET(syslog-ng_public_ip:6000)'
[2020-10-05T22:43:02.763073] Outgoing message; message='<85>1 2020-10-05T20:43:01+00:00 log_client_ip IPF 18060 IHE+RFC-3881 - <?xml version="1.0" encoding="UTF-8"?>XML_Payload
[2020-10-05T22:43:02.763290] POLLERR occurred while idle; fd='13'
[2020-10-05T22:43:02.763360] Syslog connection broken; fd='13', server='AF_INET(remote_dst_ip:dst_port)', time_reopen='5'
```

After 5 seconds the Syslog-NG server reconnects to the destination:
```
[2020-10-05T22:43:07.770322] Syslog connection established; fd='13', server='AF_INET(remote_dst_ip:dst_port)', local='AF_INET(syslog-ng_internal_ip:0)'
```

If i send a new message to the source again, log looks like this:
```
[2020-10-05T22:43:20.305833] Syslog connection accepted; fd='15', client='AF_INET(log_client_ip:10500)', local='AF_INET(syslog-ng_public_ip:6000)'
[2020-10-05T22:43:20.578296] Incoming log entry; line='<85>1 2020-10-05T20:43:19.550Z log_client_hostname IPF 20572 IHE+RFC-3881 - <?xml version="1.0" encoding="UTF-8"?>XML_Payload
[2020-10-05T22:43:20.581092] Syslog connection closed; fd='15', client='AF_INET(log_client_ip:10500)', local='AF_INET(syslog-ng_public_ip:6000)'
```

There is no outgoing message anymore, but no error as well.

Shutting down Syslog-NG with Ctrl+C i see the following:
```
^C[2020-10-05T22:55:34.484511] Running application hooks; hook='3'
[2020-10-05T22:55:34.484532] syslog-ng shutting down; version='3.23.1'
[2020-10-05T22:55:34.584616] Outgoing message; message='<85>1 2020-10-05T20:43:19+00:00 log_client_ip IPF 20572 IHE+RFC-3881 - <?xml version="1.0" encoding="UTF-8"?>XML_Payload
[2020-10-05T22:55:34.584924] I/O error occurred while writing; fd='13', error='Success (0)'
[2020-10-05T22:55:34.584945] Can't send the message rewind backlog; message='<85>1 2020-10-05T20:43:19+00:00 log_client_ip IPF 20572 IHE+RFC-3881 - <?xml version="1.0" encoding="UTF-8"?>XML_Payload
[2020-10-05T22:55:34.585026] Disk-buffer state saved; filename='/var/lib/syslog-ng/syslog-ng-00003.qf', qout_length='1', qbacklog_length='0', qoverflow_length='0', qdisk_length='0'
[2020-10-05T22:55:34.586035] Running application hooks; hook='4'
```

After starting the server again, i see the same message as when i sent the first message at the beginning of my post (Establishing connection to destination, TLS handshake, outgoing message). And i also successfully receive the message on the backend when it is resent after the server restart.
But there is an error "I/O error occurred while writing; fd='13', error='Success (0)'" which looks like an I/O error happend somewhere.

For me it looks like the server won't do a new TLS handshake after the reconnection.

I also tcpdump'ed the connection, here is a rough sequence.
This is when the destination whants to close the connection because of the 30 second timeout.
```
destination_host --> Syslog-NG_server [PSH, ACK]
destination_host --> Syslog-NG_server [FIN, ACK]
destination_host --> Syslog-NG_server [TCP Retransmission]
Syslog-NG_server --> destination_host [ACK]
```

Then i send a new message and for me it looks like the Syslog-NG server is just trying to send the data and then gets the connection reset:
```
Syslog-NG_server --> destination_host [PSH, ACK]
destination_host --> Syslog-NG_server [RST]
Syslog-NG_server --> destination_host [ACK]
destination_host --> Syslog-NG_server [RST]
Syslog-NG_server --> destination_host [ACK]
destination_host --> Syslog-NG_server [RST]
```

Then after 5 seconds of wait time for the reconnect I only see e normal SYN-ACK sqeuence.
```
Syslog-NG_server --> destination_host [SYN]
destination_host --> Syslog-NG_server [SYN, ACK]
Syslog-NG_server --> destination_host [SYN]
```

After that the server isn't sending message to the destination anymore (only after it is restarted).

Sorry for this long post, but it is really complicated to explain the problem.

I really have no idea what could be the problem. Therefore any help would be appreciated.
Willcey
[customer_test.txt](https://github.com/syslog-ng/syslog-ng/files/5330391/customer_test.txt)

**Comment by gaborznagy** [comment]
Hi @willcey ,

No need to apologize for a long post, thanks for the thorough analysis before opening an issue!

Just to confirm, do you see this error even after syslog-ng restart?
>But there is an error "I/O error occurred while writing; fd='13', error='Success (0)'" which looks like an I/O error happend somewhere.

Unfortunately, I don't have any tips at the moment, but I'll look into it deeper and give an update once I've found something.
I've looked around in the code and based on my findings I can say that the TLS handshake is always done (transparently) by calling `SSL_write()`.
If you see `Outgoing message;` message, that's a good indication that TLS handshake will be done as well, although you pointed out that you don't see them after the server reconnected to the remote_dest...


I'll try to reproduce it. If you can think of any additional detail that could help, please share it.
For example, how big are the log messages?
I see examples in your internal log that you are sending XML payloads, right?

**Comment by willcey** [comment]
Hi @gaborznagy 

> Just to confirm, do you see this error even after syslog-ng restart?

I only see the error when shutting down Syslog-NG using Ctrl+C. When I start it again, i see "outgoing message", then the "Can't send the message rewind backlog;", then the certificate validation, and then again "outgoing message" when it get sent.

Here is the log output of the restart:
```
[2020-10-12T12:41:47.909544] Finishing include; filename='/etc/syslog-ng/conf.d/customer_test.conf', depth='1'
[2020-10-12T12:41:47.910842] Accepting connections; addr='AF_INET(syslog-ng_public_ip:6000)'
[2020-10-12T12:41:47.910938] Reliable disk-buffer state loaded; filename='/var/lib/syslog-ng/syslog-ng-00009.rqf', queue_length='1', size='3717'
[2020-10-12T12:41:47.910949] Reliable disk-buffer internal state; filename='/var/lib/syslog-ng/syslog-ng-00009.rqf', backlog_head='4096', read_head='4096', write_head='7813', backlog_len='0'
[2020-10-12T12:41:47.913035] Running application hooks; hook='1'
[2020-10-12T12:41:47.913047] Running application hooks; hook='6'
[2020-10-12T12:41:47.913101] syslog-ng starting up; version='3.23.1'
[2020-10-12T12:41:47.913109] Running application hooks; hook='2'
[2020-10-12T12:41:47.913134] Syslog connection established; fd='13', server='AF_INET(remote_dst_ip:dst_port)', local='AF_INET(syslog-ng_internal_ip:0)'
[2020-10-12T12:41:47.913327] Outgoing message; message='<85>1 2020-10-12T10:41:06+00:00 log_client_ip IPF 20960 IHE+RFC-3881 - XML_Payload
[2020-10-12T12:41:47.913497] Can't send the message rewind backlog; message='<85>1 2020-10-12T10:41:06+00:00 log_client_ip IPF 20960 IHE+RFC-3881 - XML_Payoad
[2020-10-12T12:41:47.924219] Certificate validation progress; subject='CN=Root CA, OU=, O=Organization, C=CH', issuer='CN=Root CA, OU=, O=Organization, C=CH'
[2020-10-12T12:41:47.924328] Certificate validation progress; subject='CN=Intermediate CA, OU=, O=Organization, C=CH', issuer='CN=Root CA, OU=, O=Organization, C=CH'
[2020-10-12T12:41:47.924379] Certificate validation progress; subject='CN=remote_dst_host, OU=, O=Organization, L=Location, ST=State, C=CH', issuer='CN=Intermediate CA, OU=, O=Organization, C=CH'
[2020-10-12T12:41:47.924388] Certificate subject matches configured hostname; hostname='remote_dst_host', certificate='remote_dst_host'
[2020-10-12T12:41:47.934221] Outgoing message; message='<85>1 2020-10-12T10:41:06+00:00 log_client_ip IPF 20960 IHE+RFC-3881 - XML_Payload
```

> If you see Outgoing message; message, that's a good indication that TLS handshake will be done as well, although you pointed out that you don't see them after the server reconnected to the remote_dest...

Thats correct, i see them as long as Syslog-NG is "connected" to the backend. After 30 seconds of inactivity the backend closes the connection. Then Syslog-NG reconnects, but if i am sending new messages i only see "incoming message" but no "outgoing message".

> For example, how big are the log messages?

The log file has a size of about 3398 bytes.

> I see examples in your internal log that you are sending XML payloads, right?

Yes i am sending an XML payload. But that should not be an issue, because the message is able to reach the backend under some circumstances. Thats why i am really at the end of any ideas.

**Comment by gaborznagy** [comment]
Thanks for the update and to check it without TLS! So now we know it must be around TLS.

Sorry, I didn't have time to dive deep into this until now, but now I've started to try reproduce this.
I've just started, but already found something else around message rewinding... gonna check that too.

I'm using `openssl s_server` for testing and with that I was terminating the connection (with FIN packets), so syslog-ng on the next message sending detected that the connection is broken.
I can spend more time on this issue, so I'll keep going with connection reset too.

**Comment by gaborznagy** [comment]
I've just wanted to give an update: Unfortunately, I cannot reproduce this issue so far.
I do see the same `POLLERR occurred while idle`, but the connection doen't stuck in `CLOSE_WAIT` state in my setup.

My reproduction steps in short: I'm just using openssl s_server on the remote server and after a successful connection and couple of messages sent from syslog-ng, I just close the openssl s_server on the remote host, resulting in a FIN-ACK - ACK packet exchange. See tcpdump:
```
11:49:36.888060 IP remote_host > syslog-ng: Flags [F.], seq 3648, ack 495, win 235, options [nop,nop,TS val 1928821889 ecr 742637706], length 0
11:49:36.928947 IP syslog-ng > remote-host: Flags [.], ack 3649, win 503, options [nop,nop,TS val 742661328 ecr 1928821889], length 0
```
After this I do see the connection in `CLOSE_WAIT` state, but syslog-ng hasn't sent any messages yet to the remote host, see lsof output:
```
syslog-ng 19718 gaborznagy   11u     IPv4             397387      0t0      TCP localhost:1234 (LISTEN)
syslog-ng 19718 gaborznagy   14u     IPv4             397388      0t0      TCP syslog-ng->remote_host (CLOSE_WAIT)
```
When I send a message to the remote host, the aforementioned `POLLERR` can be seen:
(Note, that in my logs the `Closing log transport` message can be seen, while in your logs it's not there, the reason is that this message is on trace level.)
```
[2020-10-19T14:13:41.799187] Outgoing message; message='<13>1 2020-10-19T14:13:41+02:00 gaborzbot - - - - foo5\x0a'
[2020-10-19T14:13:41.801257] POLLERR occurred while idle; fd='14'
[2020-10-19T14:13:41.801406] Closing log transport fd; fd='14'
[2020-10-19T14:13:41.801462] Syslog connection broken; fd='14', server='AF_INET(remote_host)', time_reopen='5'
[2020-10-19T14:13:46.811140] Syslog connection failed; fd='14', server='AF_INET(remote_host)', error='Connection refused (111)', time_reopen='5'
```
However, after I restart the openssl s_server application on the remote host the connection is re-established:
```
syslog-ng 19828 gaborznagy   11u     IPv4             395074      0t0      TCP localhost:1234 (LISTEN)
syslog-ng 19828 gaborznagy   14u     IPv4             395863      0t0      TCP syslog-ng->remote_host (ESTABLISHED)
```
And on sending messages from syslog-ng, they successfully arrive.

I use the same config options what you do, I've tried it with different openssl version (1.0.2g and 1.1.1f) too.
I've used syslog-ng version `3.23.1` and the latest `3.29.1` too, without any success.
I've even modified the code and commented out the socket closing code parts, which resulted in that the fd's are leaked, but the connection is still closed and later can be established after time_reopen. (I was looking for an error in the code, when we don't close the connection in some cases, but since I've removed all close()/shutdown() calls and the connection is still closed, I've dropped this idea).

Maybe I miss something, can you give us more detail about your network setup, please? Is there any firewall, loadbalancer between syslog-ng and the remote host? I've seen you configured `localip()`. Can you tell how it is configured, please?

**Comment by gaborznagy** [comment]
@willcey  Have you considered a workaround for the time? Would TCP keep-alive packets keep the connection open, or the application would still close the connection? You can set it from syslog-ng with `tcp-keepalive-time(), tcp-keepalive-intvl and tcp-keepalive-probes()` options.

Also just for confirmation, can you run a `netstat` on your remote destination, please?
I guess it's its waiting for the FIN packet, right?
I see this before syslog-ng closes the conenction, but then as I've said, in my setup syslog-ng closes the connection.
```
tcp        0      0 remote_host:4444       syslog-ng:46175       FIN_WAIT2
```

I still wonder why does reconnect work for you with plain TCP and not with TLS....

**Comment by willcey** [comment]
Hi @gaborznagy 

I will try the one with openssl s_server. Cloud you maybe send me the command line of openssl s_server?

On the backend i see the following netstat
´´´
tcp6       0      0 remote_dst_ip:dst_port      syslog-ng_internal_ip:35555      FIN_WAIT2
´´´

While on the syslog-ng server i see the following netstat.
´´´
tcp       32      0 syslog-ng_internal_ip:43111      remote_dst_ip:dst_port      CLOSE_WAIT  472249/syslog-ng
´´´

I made another tcpdump an saw that when i shutdown the syslog-ng server a "Client Hello" is sent to the backend using TLS 1.0 before it stops. All other traffic is is TLS 1,2. Maybe this could be a hint?

**Comment by gaborznagy** [comment]
The openssl s_server command I used: `openssl s_server -key <server.key> -cert <server.crt> -CAfile <ca.crt>  -accept 4444 -state -debug` (I was using self-signed certificates for testing).
I've played with some openssl options like `-no_ssl3 -no_tls1 -no_tls1_1` but they didn't made any difference.
For reproduction just simply close the openssl application, when syslog-ng doesn't send any logs, then wait for the TCP connection to be re-established, then send a message to syslog-ng to see if the TLS handshake happens or not.

I'll recheck the shutdown part you mentioned, but I don't know if it's a hint.

What do you think about the proposed workarounds?

**Comment by willcey** [comment]
Sorry for my late answer.

The workaround with tcp-keepalive did not work for me. But could you maybe give me a suggestion how I should configure them. The destination closes the connection after 30 seconds without message.
At the moment I set the following (maybe this is wrong?):
tcp-keepalive-intvl(5)
tcp-keepalive-time(5)
tcp-keepalive-probes(1)

I was able to run the destination against openssl s_server and it works also for me as you already said. I will try to get in touch with the manufacturer of the destination software.

Another question: is it possible to set syslog-ng to open a new connection for every message that needs to be sent to the destination. I know this is not recommended because of the tcp protocol overhead, but is there a configuration switch for doing this?

**Comment by willcey** [comment]
Hello

We did not find any solution to this with our other software provider. The only way it worked was to send mark-messages every couple of seconds which filled up the log file of our backend.

So we have choosen to put s_tunnel in between. Now our setup looks like this:
Customer --TLS--> Syslog-NG --unencrypted--> s_tunnel --TLS--> Backend

This seems to work, but we had to put the unencrypted traffic between Syslog-NG and s_tunnel. We also tried the following:
Customer --TLS--> Syslog-NG --TLS--> s_tunnel --unencrypted--> Backend

But this gave us the same error like when Syslog-NG is talking directly to the backend.

Maybe this information is helpful.
